src/swt.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2019 King's College London.
+// Created by the Software Development Team <http://soft-dev.org/>.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use super::{MirTrace, ThreadTracer};
+/// Software tracing via ykrustc.
+use core::yk_swt::{self, MirLoc};
+use libc;
+use std::ops::Drop;
+
+/// A trace collected via software tracing.
+/// Since the trace is a heap-allocated C buffer, we represent it as a pointer and a length.
+struct SWTMirTrace {
+    buf: *mut MirLoc,
+    len: usize
+}
+
+impl MirTrace for SWTMirTrace {
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn loc(&self, idx: usize) -> &MirLoc {
+        if idx > isize::max_value() as usize {
```

Comment:
Why are we checking against `isize::max_value`? Is it because we're using `offset` below? If so, `add` negates the need for this check (assuming, as I think we must, that `ptr + self.len` is in bounds).

Reply:
> Is it because we're using offset below?

Yeah, offset accepts an isize so that you can bump pointers both back and forward:
```
pub unsafe fn offset(self, count: isize) -> *const T
```

> pub unsafe fn offset(self, count: isize) -> *const T

Ah! I did not know about `Ptr::add`! Marvellous. Will fix.